Azure Logs documentation update

zmoog opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
-->

## What does this PR do?

<!-- Mandatory
Explain here the changes you made on the PR.
-->

Expand Azure Logs integration docs to make it easier for users to set it up.

It addresses issues **2** and **5** from #4169.

~~**Disclaimer**: this is an early draft to share the change with the cloud monitoring team members and docs wizards.~~

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [ ] ~~I have verified that all data streams collect metrics or logs.~~
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

<!-- Recommended

## Author's Checklist
Add a checklist of things that are required to be reviewed in order to have the PR approved
- [ ]
-->

<!-- Recommended
## How to test this PR locally
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Supersedes #123
-->
- Relates #4169 
- Relates #4210 

## Screenshots

<!-- Optional
Add here screenshots presenting:
- Kibana UI forms presenting configuration options exposed by the integration
- dashboards with collected metrics or logs
-->

![CleanShot 2022-09-30 at 13 14 41@2x](https://user-images.githubusercontent.com/25941/193258071-e106c26b-3fe8-48ea-a84d-8757f2db4931.png)

elasticmachine commented:
### :rocket: Benchmarks report




To see the full report comment with `/test benchmark fullreport`

<!--COMMENT_GENERATED_WITH_ID_benchmark-report.md-->

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-4300/detail/PR-4300/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-4300/detail/PR-4300/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-4300/detail/PR-4300/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-4300/detail/PR-4300/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://integrations_4300.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2022-09-26T10:14:13.353Z&rangeTo=2022-09-26T10:34:13.353Z&transactionName=BUILD+Ingest-manager%2Fintegrations%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=9d157455fe791eb6bb1ad44399939eac&transactionId=083f6dadbbedac9c)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-09-26T10:24:13.353+0000


* Duration: 15 min 6 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 119  |
| Skipped      | 0 |
| Total        | 119   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 100.0% (`9/9`) | :green_heart: 
Files | 85.0% (`17/20`) | :-1: -12.418
Classes | 85.0% (`17/20`) | :-1: -12.418
Methods | 82.143% (`138/168`) | :-1: -7.796
Lines | 85.578% (`2504/2926`) | :-1: -5.787
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

colleenmcginnis commented:
👋  @zmoog I like the direction you're taking here!

One thing that isn't clear (yet) in this draft is the relationship between this [Azure Logs](https://docs.elastic.co/integrations/azure) integration and the individual [Active Directory](https://docs.elastic.co/integrations/azure/adlogs), [Activity logs](https://docs.elastic.co/integrations/azure/activitylogs), [Firewall logs](https://docs.elastic.co/integrations/azure/firewall_logs), [Platform logs](https://docs.elastic.co/integrations/azure/platformlogs), and [Spring Cloud logs](https://docs.elastic.co/integrations/azure/springcloudlogs) integrations. Are they intended to be used together? Or should a user choose _either_ the Azure Logs integration _or_ one or more of the other integrations for collecting logs? In what situations should you choose each approach?

zmoog commented:
> 👋 @zmoog I like the direction you're taking here!

That's good! So I'll move ahead exploring how to tackle the Setup section.

> One thing that isn't clear (yet) in this draft is the relationship between this [Azure Logs](https://docs.elastic.co/integrations/azure) integration and the individual [Active Directory](https://docs.elastic.co/integrations/azure/adlogs), [Activity logs](https://docs.elastic.co/integrations/azure/activitylogs), [Firewall logs](https://docs.elastic.co/integrations/azure/firewall_logs), [Platform logs](https://docs.elastic.co/integrations/azure/platformlogs), and [Spring Cloud logs](https://docs.elastic.co/integrations/azure/springcloudlogs) integrations. Are they intended to be used together? Or should a user choose _either_ the Azure Logs integration _or_ one or more of the other integrations for collecting logs? In what situations should you choose each approach?

I haven't touched the documentation of the individual integrations yet. The idea is to replicate the AWS doc approach: general and shared information in the main README.md file, and the specific information and references (fields and sample logs) in the individual .md files.

zmoog commented:
Hey @colleenmcginnis, I expanded the Setup section drafting the kind of content that users probably need to set up Azure Logs. I also moved some details from the Requirements section to the Setup one.

Let me know what you think approach!

The next step is revise the Settings section and start working on the individual integration pages, and move the references there from the README.

zmoog commented:
@colleenmcginnis, I reduced the scope of this PR a little; I plan a series of small, more focused PRs next.

In this PR, I focus on addressing issues 2 and 5 from #4169 (address the missing information about event hubs and the ambiguity between event hub namespace and event hub.). 

I also removed the field details and sample events from the Reference section on the main page, similar to the AWS integration doc.

alaudazzi commented on packages/azure/_dev/build/docs/README.md:
```diff
@@ -1,38 +1,207 @@
+
+### Create a Storage account container
+
+The Storage account container stores the consumer group info (state, position, or offset). This allow the Elastic Agent to resume from the last processed logs after a restart, or share the load between multiple Elastic Agents.
```
```suggestion
The Storage account container stores the consumer group info (state, position, or offset). This allows the Elastic Agent to resume from the last processed logs after a restart, or share the load between multiple Elastic Agents.
```

alaudazzi commented on packages/azure/_dev/build/docs/README.md:
```diff
@@ -1,38 +1,207 @@
-and reference data when troubleshooting an issue.
+The Azure Logs integration retrieves Activity, Active Directory (Sign-in, Audit, Identity Protection, Provisioning), Platform, and Spring Cloud data from [Azure](https://docs.microsoft.com/en-us/azure/?product=popular).
+
+Use the Azure Logs integration to collect logs from Azure service. Then visualize that data in Kibana, create alerts to notify you if something goes wrong, and reference data when troubleshooting an issue.
```
```suggestion
Use the Azure Logs integration to collect logs from Azure service. You can then visualize that data in Kibana, create alerts to notify you if something goes wrong, and reference data when troubleshooting an issue.
```

alaudazzi commented on packages/azure/_dev/build/docs/README.md:
```diff
@@ -1,38 +1,207 @@
+**Logs** help you keep a record of events that happen on your Azure account.
+Log data streams collected by the Azure Logs integration include Activity, Platform, Active Directory (Sign-in, Audit, Identity Protection, Provisioning), and Spring Cloud logs.
+
 See more details in the [Logs reference](#logs-reference).
```
```suggestion
Check the [Logs reference](#logs-reference) for more details.
```
Replace visual clues with more accessible content. [Accessibility Tips and Tricks.](https://docs.google.com/document/d/1IGuMf16-cwxx8BX1Z9PWsFWsTPxicwTkX_r7UOhjd4E/edit)

alaudazzi commented on packages/azure/_dev/build/docs/README.md:
```diff
@@ -1,38 +1,207 @@
+
+* One or more **Diagnostic setting** to export logs from Azure services to Event Hubs.
+* One or more **Event Hub** to store in-flight logs exported by Azure services and make them available to Elastic Agent.
+* One **Storage Account Container** to store information about logs consumed by the Elastic Agent
```
```suggestion
* One **Storage Account Container** to store information about logs consumed by the Elastic Agent.
```

alaudazzi commented on packages/azure/_dev/build/docs/README.md:
```diff
@@ -1,38 +1,207 @@
+* One or more **Event Hub** to store in-flight logs exported by Azure services and make them available to Elastic Agent.
+* One **Storage Account Container** to store information about logs consumed by the Elastic Agent
+
+### Diagnostic setting
```
```suggestion
### Diagnostic settings
```

alaudazzi commented on packages/azure/_dev/build/docs/README.md:
```diff
@@ -1,38 +1,207 @@
+
+### Diagnostic setting
+
+Azure Diagnostic settings allow users to export metrics and logs from a **source** service or resource to one **destination** for analysis and long term storage.
```
```suggestion
Azure Diagnostic settings allow you to export metrics and logs from a **source** service, or resource, to one **destination** for analysis and long-term storage.
```

alaudazzi commented on packages/azure/_dev/build/docs/README.md:
```diff
@@ -1,38 +1,207 @@
+
+### Event Hub
+
+[Azure Event Hubs](https://learn.microsoft.com/en-us/azure/event-hubs/event-hubs-about) is a data streaming platform and event ingestion service. I can receive and temporary store millions of events.
```
```suggestion
[Azure Event Hubs](https://learn.microsoft.com/en-us/azure/event-hubs/event-hubs-about) is a data streaming platform and event ingestion service. It can receive and temporary store millions of events.
```

alaudazzi commented on packages/azure/_dev/build/docs/README.md:
```diff
@@ -1,38 +1,207 @@
+
+The Azure Logs integration uses the Event Hubs service to receive and store logs exported by a Diagnostic settings and make them available to Elastic Agent.
+
+To learn more about Event Hubs, you can read the in-depth document [Features and terminology in Azure Event Hubs](https://learn.microsoft.com/en-us/azure/event-hubs/event-hubs-features).
```
```suggestion
To learn more about Event Hubs, refer to [Features and terminology in Azure Event Hubs](https://learn.microsoft.com/en-us/azure/event-hubs/event-hubs-features).
```

alaudazzi commented on packages/azure/_dev/build/docs/README.md:
```diff
@@ -1,38 +1,207 @@
+
+### Storage account container
+
+The Storage account is a versatile Azure service that allows users to store data in various storage types, including blobs, file shares, queues, tables, and disks.
```
```suggestion
The Storage account is a versatile Azure service that allows you to store data in various storage types, including blobs, file shares, queues, tables, and disks.
```

alaudazzi commented on packages/azure/_dev/build/docs/README.md:
```diff
@@ -1,38 +1,207 @@
+* Create an Event Hubs namespace.
+* Create an event hub.
+
+For a detailed step-by-step guide, please follow the instructions at [Quickstart: Create an event hub using Azure portal](https://learn.microsoft.com/en-us/azure/event-hubs/event-hubs-create).
```
```suggestion
For a detailed step-by-step guide, check the quickstart [Create an event hub using Azure portal](https://learn.microsoft.com/en-us/azure/event-hubs/event-hubs-create).
```

alaudazzi commented on packages/azure/_dev/build/docs/README.md:
```diff
@@ -1,38 +1,207 @@
+
+#### Connection String
+
+The Elastic Agent requries a Connection String to access the Event Hub and fetch the exported logs. It contains details about the Event Hub used and the credentials required to access it.
```
```suggestion
The Elastic Agent requires a Connection String to access the Event Hub and fetch the exported logs. It contains details about the Event Hub used and the credentials required to access it.
```

alaudazzi commented on packages/azure/_dev/build/docs/README.md:
```diff
@@ -1,38 +1,207 @@
+The Diagnostic settings export the logs from Azure services to a destination. The Azure Logs integration uses an event hub as the destination for the logs.
+
+1. Locate the Diagnostic settings for the service (for example, Azure Active Directory).
+1. Select Diagnostic settings in the **Monitoring** section of the service. Please note that different services may place the diagnostic settings in different positions.
```
```suggestion
1. Select Diagnostic settings in the **Monitoring** section of the service. Note that different services may place the diagnostic settings in different positions.
```

alaudazzi commented on packages/azure/_dev/build/docs/README.md:
```diff
@@ -1,38 +1,207 @@
+
+Take note of the **Storage account name**, which you will use later when specifying the **storage_account** in the integration settings.
+
+When the new Storage account is ready, we can look for the access keys:
```
```suggestion
When the new Storage account is ready, you can look for the access keys:
```

alaudazzi reviewed (approved):
Left a few editing suggestions, otherwise LGTM.

alaudazzi commented:
@zmoog 
Great PR 🥇 Thank you for reworking these instructions!  
As we discussed yesterday, reducing the scope of this PR and address the remaining items from https://github.com/elastic/integrations/issues/4169 makes totally sense.

zmoog commented on packages/azure/_dev/build/docs/README.md:
```diff
@@ -1,169 +1,282 @@
+Like all other clients, Elastic Agent should specify a consumer group to access the event hub.
 
-## Logs reference
+A Consumer Group is a view (state, position, or offset) of an entire event hub. Consumer groups enable multiple agents to each have a separate view of the event stream, and to read the logs independently at their own pace and with their own offsets.
```
We need to go one level deeper to understand what's going on.

The current integration structure forces the same event hub and consumer group to be shared across all enabled data streams.

Every enabled data stream spawns an azureeventhub input that connects to the same event hub and uses the same consumer group name:

```text
                                                     
                   ┌────────────────┐                
                   │     adlogs     │                
                   │ <<event hub>>  │                
                   └────────────────┘                
                            │                        
                            │                        
                            │                        
             ┌──────────────┼──────────────┐         
         $Default       $Default       $Default      
             │              │              │         
  ┌──────────┼──────────────┼──────────────┼────────┐
  │          ▼              ▼              ▼        │
  │   ┌────────────┐ ┌────────────┐ ┌────────────┐  │
  │   │   signin   │ │   audit    │ │  activity  │  │
  │   │ <<input>>  │ │ <<input>>  │ │ <<input>>  │  │
  │   └────────────┘ └────────────┘ └────────────┘  │
  │          │              │              │        │
  └─Filebeat─┼──────────────┼──────────────┼────────┘
             │              │              │         
             │              │              │         
             │     consumer group info     │         
             │    (state, position, or     │         
             │           offset)           │         
             │              │              │         
             │              │              │         
  ┌──────────┼──────────────┼──────────────┼────────┐
  │          ▼              ▼              ▼        │
  │   ┌────────────┐  ┌──────────┐   ┌──────────┐   │
  │   │   signin   │  │  audit   │   │  audit   │   │
  │   │  <<blob>>  │  │ <<blob>> │   │ <<blob>> │   │
  │   └────────────┘  └──────────┘   └──────────┘   │
  │                                                 │
  └─storage account container───────────────────────┘
```

But, since each input stores the consumer group info (state, position, or offset) on a different blob, it's like each uses a dedicated consumer group. Even if they use the same consume group name.

That's why each data stream connected to the same data stream receives a copy of each message.

zmoog replied:
If you assign the same agent policy to multiple Elastic Agents, they end up using the same blob and meet the goal of sharing the load:

```text
               ┌────────────────┐                  
               │     adlogs     │                  
               │ <<event hub>>  │                  
               └────────────────┘                  
                        │                          
                        │                          
                        │                          
           ┌────────────┴───────────┐              
       $Default                 $Default           
           │                        │              
┌──────────┼──────────┐  ┌──────────┼──────────┐   
│          ▼          │  │          ▼          │   
│   ┌────────────┐    │  │   ┌────────────┐    │   
│   │   signin   │    │  │   │   signin   │    │   
│   │ <<input>>  │    │  │   │ <<input>>  │    │   
│   └────────────┘    │  │   └────────────┘    │   
│          │          │  │          │          │   
└─Filebeat─┼──────────┘  └─Filebeat─┼──────────┘   
           │                        │              
           │                        │              
           │                        │              
           │ consumer group info    │              
           │(state, position, or    │              
           │       offset)          │              
           │                        │              
┌──────────┼────────────────────────┼─────────────┐
│          │                        │             │
│          │     ┌────────────┐     │             │
│          │     │   signin   │     │             │
│          └────▶│  <<blob>>  │◀────┘             │
│                └────────────┘                   │
│                                                 │
└─storage account container───────────────────────┘
```

In this example, both Filebeat instances will use the same blob. The blob name is identified by name using the following pattern: `filebeat-signinlogs-{{eventhub}}`.

zmoog replied:
@lucabelluccini, what do you think is the best way to describe the role of the consumer group to a user who wants to set up an integration?

zmoog replied:
We agreed to make the following changes:

- add a paragraph that describes that consumer groups allow Agent to collaborate and scale the throughput
- underline that having an event hub per log type is a good thing

zmoog commented on packages/azure/_dev/build/docs/README.md:
```diff
@@ -1,169 +1,282 @@
+A Consumer Group is a view (state, position, or offset) of an entire event hub. Consumer groups enable multiple agents to each have a separate view of the event stream, and to read the logs independently at their own pace and with their own offsets.
 
-### Activity logs
+In most cases, you can use the default value of `$Default`.
```
Yep. I am not an expert here, but I think having a consumer group named after the log type it serves would help. I will update the doc accordingly.

zmoog commented on packages/azure/_dev/build/docs/README.md:
```diff
@@ -1,169 +1,282 @@
+### Create a Diagnostic settings
 
-{{event "signinlogs"}}
+The Diagnostic settings export the logs from Azure services to a destination. The Azure Logs integration uses the Event Hub service as the destination for the logs.
```
I like it!

zmoog commented on packages/azure/_dev/build/docs/README.md:
```diff
@@ -1,169 +1,282 @@
+#### How many event hubs?
 
-Examples:
+Elastic recommends creating one event hub for each Azure service you collect data from. For example, if you plan to collect Azure Active Directory (Azure AD) logs and Activity logs, create two event hubs: one for Azure AD and one for Activity logs.
```
Some integrations are designed to filter incoming logs and drop those that are not on the supported log categories list.

For example, both sign-in and audit logs have a processor in their ingest pipelines that drop logs that don't belong to the supported log categories. So you can enable all the Azure AD logs, and each one will ingest only the right log messages. The price is inefficiencies due to all data streams receiving the same messages: dropping a message is not completely free.

Other integrations, like the generic Event Hub integration, can ingest any log category but it indexes only the common fields. So enabling a generic Event Hub integration alongside others is not recommended and can lead to indexing the same log multiple times, in different data streams, and with different field mapping.

@lucabelluccini, do you think we need to elaborate a little instead of just stating "Elastic recommends creating one event hub"?

lucabelluccini replied:
> For example, both sign-in and audit logs have a processor in their ingest pipelines that drop logs that don't belong to the supported log categories.

If the drop is in the ingest pipeline on Elasticsearch side, leaving all the types of logs in the Event Log settings is a waste of network and ingest pipeline resources on the ingest nodes, as we're basically multiplying N times (N = the type of logs/data streams left enabled)

E.g. I collect only the AD logs and I route them to 1 event hub. If I configure the Azure Logs integration as per defaults (all options enabled) as below, the amount of data transferred to ES (for then being thrown away except for the AD) is huge...

<img width="797" alt="image" src="https://user-images.githubusercontent.com/87379/193093356-74af3a02-1ac8-47bc-8fbd-6efc0ab41ff8.png">

zmoog replied:
Yep. That's why I want to update this document so badly.

I also want to open a different PR to change the default value from enabled to disabled (the new data streams added to Azure AD logs are disabled by default). 

The Azure module for Filebeat also [recommends](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-module-azure.html) one even hub per log type:
>  It is recommended to use a separate eventhub for each log type as the field mappings of each log type are different.

The reason is different but still valid.

zmoog replied:
We agreed to make the following changes:

- Add a disclaimer at the top of the README.md document suggesting installing the "individual" integrations.
- We "strongly recommend"

zmoog commented on packages/azure/_dev/build/docs/README.md:
```diff
@@ -1,169 +1,282 @@
-This setting can also be used to define your own endpoints, like for hybrid cloud models.
+It is not recommended to use the same event hub for multiple integrations.
+
+For high-volume deployments, we recommend one event hub for each data stream.
```
The goal was to introduce the idea that high-vol data streams may require an additional work.

A good trade-off for most users is one event hub for all Azure AD logs (made of four data streams now), but if you have a substantial Active Directory deployment, you may consider moving to one event hub for each data stream.

zmoog commented:
Hey @alaudazzi @lucabelluccini, I pushed an update that addresses the topic we discussed earlier today.

Let me know what you think! I'm more than happy to clarify, expand or fix errors.

aspacca commented on packages/azure/_dev/build/docs/README.md:
```diff
@@ -1,169 +1,314 @@
+
+The [Storage account](https://learn.microsoft.com/en-us/azure/storage/common/storage-account-overview) is a versatile Azure service that allows you to store data in various storage types, including blobs, file shares, queues, tables, and disks.
+
+The Azure Logs integration uses a Storage account container to store and share information about the Consumer Group (state, position, or offset). Sharing such information allows the Elastic Agents assigned to the same agent policy to work together on the logs processing to increase ingestion throughput if required.
```
as a non-native english speaker I miss the detail of having "Agents" at plural: I know the storage account is used for horizontal scaling so I got it anyway, maybe having something like "Sharing such information allows *multiple* Elastic Agents assigned to the same agent policy to work together on the logs processing to increase ingestion throughput if required" make it more evident

zmoog replied:
Oh, this is a good point! I had a vague feeling that "Agents" was a sort of a smell, and you put it in focus. 

Updating it with your version!

aspacca commented on packages/azure/_dev/build/docs/README.md:
```diff
@@ -1,169 +1,314 @@
+
+A Consumer Group is a view (state, position, or offset) of an entire event hub. Consumer groups enable multiple agents to each have a separate view of the event stream, and to read the logs independently at their own pace and with their own offsets.
+
+Consumer groups allow the Elastic Agents assigned to the same agent policy to work together on the logs processing to increase ingestion throughput if required.
```
same as above :)

aspacca replied:
I find also that mentioning the concept of “horizontal scaling" might increase clarity. but it could be a subjective opinion

zmoog replied:
Yeah, I agree, and I'm also stealing the "horizontal scaling" thing 😇

tommyers-elastic commented on packages/azure/changelog.yml:
```diff
@@ -1,3 +1,8 @@
+  changes:
+    - description: "Update Azure Logs documentation"
+      type: enhancement
+      link: https://github.com/elastic/integrations/pull/9999 # FIXME: link to PR
```
```suggestion
      link: https://github.com/elastic/integrations/pull/4300
```

zmoog replied:
Fixed! Thank you!
